### PR TITLE
[SHELL32][SDK] Implement CopyStreamUI

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Implementing missing features...
JIRA issue: [CORE-19278](https://jira.reactos.org/browse/CORE-19278)

## Proposed changes

- Modify `shell32.spec`.
- Implement `CopyStreamUI` function.
- Add `CopyStreamUI` prototype to `<undocshell.h>`.

## TODO

- [x] Do build.

### CODE REVIEW

File: dll/win32/shell32/utils.cpp [hunk 1]
```diff
@@ -540,6 +540,123 @@ LargeIntegerToString(
                          pNumberFormat, dwNumberFlags);
 }
 
+/*************************************************************************
+ *  CopyStreamUI [SHELL32.726]
+ *
+ * Copy a stream to another stream with displaying progress.
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:546]:
```suggestion
 * Copy a stream to another stream with optional progress display.
```

File: dll/win32/shell32/utils.cpp [hunk 2]
```diff
@@ -540,6 +540,123 @@ LargeIntegerToString(
...
+    STATSTG Stat;
+    BYTE szBuff[1024];
+
+    TRACE("(%p, %p, %p, %I64u)\n", pSrc, pDst, pProgress, dwlSize);
+
+    if (dwlSize == 0) // Invalid size?
+    {
+        // Get the stream size
+        ZeroMemory(&Stat, sizeof(Stat));
+        if (FAILED(pSrc->Stat(&Stat, 1)))
```
Comment (learn-more) [inline dll/win32/shell32/utils.cpp:569]: `STATFLAG_NONAME`

File: dll/win32/shell32/utils.cpp [hunk 3]
```diff
@@ -540,6 +540,123 @@ LargeIntegerToString(
...
+CopyStreamUI(
+    _In_ IStream *pSrc,
+    _Out_ IStream *pDst,
+    _Inout_opt_ IProgressDialog *pProgress,
+    _In_opt_ DWORDLONG dwlSize)
+{
+    HRESULT hr = E_FAIL;
+    DWORD cbBuff, cbRead, dwSizeToWrite;
+    DWORDLONG cbDone;
+    LPVOID pBuff;
```
Comment (learn-more) [inline dll/win32/shell32/utils.cpp:559]: Why not use a smart pointer for this?

File: dll/win32/shell32/utils.cpp [hunk 4]
```diff
@@ -540,6 +540,123 @@ LargeIntegerToString(
...
+    }
+
+    // Start reading
+    LARGE_INTEGER zero;
+    zero.QuadPart = 0;
+    pSrc->Seek(zero, 0, NULL);
+    pDst->Seek(zero, 0, NULL);
+    cbDone = 0;
+    if (pProgress) // Progress is enabled?
+        pProgress->SetProgress64(cbDone, dwlSize);
```
Comment (learn-more) [inline dll/win32/shell32/utils.cpp:613]: We cannot come here if `pProgress == null`, so why this check?

File: dll/win32/shell32/utils.cpp [hunk 5]
```diff
@@ -540,6 +540,123 @@ LargeIntegerToString(
...
 }
 
+/*************************************************************************
+ *  CopyStreamUI [SHELL32.726]
+ *
+ * Copy a stream to another stream with optional progress display.
+ */
+EXTERN_C
+HRESULT WINAPI
+CopyStreamUI(
```
Comment (learn-more) [inline dll/win32/shell32/utils.cpp:550]: Do you know an application that uses this function?

File: dll/win32/shell32/utils.cpp [hunk 6]
```diff
@@ -540,6 +540,123 @@ LargeIntegerToString(
...
+            break;
+        }
+
+        hr = pDst->Write(pBuff, dwSizeToWrite, NULL);
+        if (hr != S_OK)
+            break;
+
+        cbDone += dwSizeToWrite;
+
+        if (pProgress) // Progress is enabled?
```
Comment (learn-more) [inline dll/win32/shell32/utils.cpp:636]: There is an early bail-out if pProgress is null, why the check here?

File: dll/win32/shell32/utils.cpp (hunk 5 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp:550]: `shell32` uses this function in `CLocalCopyHelper::Download`, `CLocalCopyHelper::Upload`, and `CopyStreamToFile`.

File: dll/win32/shell32/utils.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp]: Fixed.

File: dll/win32/shell32/utils.cpp (hunk 6 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp]: Removed check.

File: dll/win32/shell32/utils.cpp [hunk 7]
```diff
@@ -540,6 +540,123 @@ LargeIntegerToString(
...
+    LARGE_INTEGER zero;
+    zero.QuadPart = 0;
+    pSrc->Seek(zero, 0, NULL);
+    pDst->Seek(zero, 0, NULL);
+    cbDone = 0;
+    if (pProgress) // Progress is enabled?
+        pProgress->SetProgress64(cbDone, dwlSize);
+
+    // Repeat reading and writing until goal
+    while (SUCCEEDED(pSrc->Read(pBuff, cbBuff, &cbRead)))
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp]: Fixed.

File: dll/win32/shell32/utils.cpp (hunk 3 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp:559]: Used `CComHeapPtr`.

File: dll/win32/shell32/utils.cpp (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp]: Fixed.

File: dll/win32/shell32/utils.cpp (hunk 5 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp:550]: `CopyStreamToFile` is referred by `CFSDropTarget::_CreatePackage`.

**Pull request merged**